sample/weighted.go

```diff
@@ -0,0 +1,98 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import "math/rand"
+
+// Weighted provides sampling without replacement from a collection of items with
+// weighted probabilites.
+type Weighted struct {
+	weights []float64
+	heap    []float64
+	src     *rand.Rand
+}
+
+// NewWeighted returns a Weighted for the weights w. Elements of w are modified
+// during sampling, so a copy should be passed if the user needs to keep the values.
+// If src is nil, rand.Rand is used as the random source.
+func NewWeighted(w []float64, src *rand.Rand) Weighted {
```

Comment:
It's especially weird to have the shared state in this case because the heap memory needs to be allocated anyway, so it's not like it changes the structure from "no allocation" to "allocation".

That said, if it is really important to save the memory for this data structure, I would be supportive of a function that explicitly asks for the shared state, something like

````
// NewWeightedShared creates a Weighted structure where the memory is provided. The data will be modified as the functions are used.
// TODO(btracey): better comment.
NewWeightedShared(weights, heap []float64)
````